packages/compass-import-export/src/import/analyze-csv-fields.ts

```diff
@@ -141,77 +137,81 @@ function pickFieldType(field: CSVField): CSVParsableFieldType {
   return field.detected;
 }
 
-export function analyzeCSVFields({
+export async function analyzeCSVFields({
   input,
   delimiter,
   abortSignal,
   progressCallback,
   ignoreEmptyStrings,
 }: AnalyzeCSVFieldsOptions): Promise<AnalyzeCSVFieldsResult> {
-  return new Promise(function (resolve, reject) {
-    const byteCounter = new ByteCounter();
-
-    const result: AnalyzeCSVFieldsResult = {
-      totalRows: 0,
-      fields: {},
-      aborted: false,
-    };
-
-    let aborted = false;
-    let headerFields: string[];
-
-    const validator = new Utf8Validator();
-
-    validator.once('error', function (err: any) {
-      reject(err);
-    });
-
-    input = input.pipe(validator).pipe(byteCounter).pipe(stripBomStream());
-
-    Papa.parse(input, {
-      delimiter,
-      header: true,
-      step: function (results: Papa.ParseStepResult<PapaRowData>, parser) {
-        if (abortSignal?.aborted && !aborted) {
-          aborted = true;
-          result.aborted = true;
-          parser.abort();
+  const byteCounter = new ByteCounter();
+
+  const result: AnalyzeCSVFieldsResult = {
+    totalRows: 0,
+    fields: {},
+    aborted: false,
+  };
+
+  let fields: CSVField[];
+
+  const parseStream = Papa.parse(Papa.NODE_STREAM_INPUT, { delimiter });
+
+  let numRows = 0;
+  const analyzeStream = new Transform({
+    objectMode: true,
+    transform: (chunk: string[], encoding, callback) => {
+      if (numRows === 0) {
+        const headerFields = chunk;
+        // There's a quirk in papaparse where it extracts header fields before
+        // it finishes auto-detecting the line endings. We could pass in a
+        // line ending that we previously detected (in guessFileType(),
+        // perhaps?) or we can just strip the extra \r from the final header
+        // name if it exists.
+        if (headerFields.length) {
+          const lastName = headerFields[headerFields.length - 1];
+          headerFields[headerFields.length - 1] = lastName.replace(/\r$/, '');
         }
 
-        if (!headerFields) {
-          headerFields = results.meta.fields ?? [];
-          // There's a quirk in papaparse where it extracts header fields before
-          // it finishes auto-detecting the line endings. We could pass in a
-          // line ending that we previously detected (in guessFileType(),
-          // perhaps?) or we can just strip the extra \r from the final header
-          // name if it exists.
-          if (headerFields.length) {
-            const lastName = headerFields[headerFields.length - 1];
-            headerFields[headerFields.length - 1] = lastName.replace(/\r$/, '');
-          }
-
-          initResultFields(result, headerFields);
-        }
-
-        addRowToResult(result, headerFields, results.data, ignoreEmptyStrings);
-
-        ++result.totalRows;
+        initResultFields(result, headerFields);
+        fields = Object.values(result.fields);
```

Comment:
This caches the result of Object.values() outside the loop. The kind of micro-optimisation I'd normally not bother with, but it does begin to matter if you're looking at millions of rows.